packages/cli/src/util/plugins.ts

```diff
@@ -0,0 +1,91 @@
+import code from '../util/output/code';
+import { getColorForPkgName } from '../util/output/color-name-cache';
+import cliPkgJson from '../util/pkg';
+import { scanParentDirs } from '@vercel/build-utils';
+
+const VERCEL_PLUGIN_PREFIX = 'vercel-plugin-';
+
+export async function loadCliPlugins(
+  cwd: string,
+  logError: (errorMessage: string) => void = console.error
+) {
+  const { packageJson } = await scanParentDirs(cwd, true);
+
+  let pluginCount = 0;
+  const preBuildPlugins = [];
+  const buildPlugins = [];
+  const devServerPlugins = [];
+  const devMiddlewarePlugins = [];
+  const deps = new Set(
+    [
+      ...Object.keys(packageJson?.dependencies || {}),
+      ...Object.keys(packageJson?.devDependencies || {}),
+      ...Object.keys(cliPkgJson.dependencies),
+    ].filter(dep => dep.startsWith(VERCEL_PLUGIN_PREFIX))
+  );
+
+  for (let dep of deps) {
+    pluginCount++;
+    const resolved = require.resolve(dep, {
+      paths: [cwd, process.cwd(), __dirname],
+    });
+    let plugin;
+    try {
+      plugin = require(resolved);
+
+      const color = getColorForPkgName(dep);
+      if (typeof plugin.preBuild === 'function') {
+        preBuildPlugins.push({
+          plugin,
+          name: dep,
+          color,
+        });
+      }
+      if (typeof plugin.build === 'function') {
+        buildPlugins.push({
+          plugin,
+          name: dep,
+          color,
+        });
+      }
+      if (typeof plugin.startDevServer === 'function') {
+        devServerPlugins.push({
+          plugin,
+          name: dep,
+          color,
+        });
+      }
+      if (typeof plugin.runDevMiddleware === 'function') {
+        devMiddlewarePlugins.push({
+          plugin,
+          name: dep,
+          color,
+        });
+      }
+    } catch (error) {
+      logError(`Failed to import ${code(dep)}`);
+      throw error;
+    }
+  }
+
+  /**
+   * This is might be wrong... should we support more than one plugin defining runDevMiddleware?
+   * Should the middleware plugin ALWAYS be added? What about when a different framework defines
+   * a runDevMiddleware plugin?
+   */
+  if (devMiddlewarePlugins.length > 1) {
```

Comment:
@leo I'm not sure about the assumption that I'm making here.

Reply:
I'm thinking that if there is a plugin defining `runDevMiddleware` in the project's package.json, we should override the one we're providing in `vercel-plugin-middleware` as they likely don't want both.

Reply:
I think it would be beneficial if multiple CLI Plugins are able to provide Middleware at the same time.

One of the big benefits of Edge Functions is that we're planning to let people run them "on top of each other", which allows for plug-and-play setups the world hasn't seen before – like one-click installation of Sentry through us automatically adding it as another Middleware, and since Middlewares can capture each other's responses, we can capture the crash stack trace of another middleware. Same for logs, etc.

Reply:
@leo Plugins would still be able to run multiple middlewares if needed, a plugin's `runDevMiddleware` would run them sequentially.  For example if next defined one, and had a filesystem like below, it'd run two middlewares for a request to `/nested` before returning the result:

```
pages/_middleware.ts
pages/nested/_middleware.ts
pages/nested/index.ts
```

BUT we probably shouldn't run `_middleware.ts` (if added to project root) via the `vercel-plugin-middleware`.

Reply:
Right, what I mean is this:

> I think it would be beneficial if multiple CLI Plugins are able to provide Middleware at the same time.

Not that one CLI Plugin would be able to run multiple Middlewares at the same time.

Basically, I'm thinking that, we don't yet know how Integrations like Sentry would be one-click-enabled, but if multiple CLI Plugins could provide Middleware, the user could provide some in `_middleware.js` and then we could auto-install `vercel-plugin-sentry`, which provides another one that runs on top, for example.

Reply:
@leo I've updated to allow for more than one again, and the dev server now runs them sequentially until it finds a result w/ `finished: true` do you have an example of a middleware intercepting another middleware's response?

I adapted from next.js's implementation originally and doesn't look like multiple middlewares there are ever passed the result of a previous middleware... maybe that wasn't the best place to be looking?

https://github.com/vercel/next.js/blob/canary/packages/next/server/next-server.ts#L671

They also only seem to run multiple matching middlewares if the `x-middleware-next` header is present:
https://github.com/vercel/next.js/blob/canary/packages/next/server/next-server.ts#L687

Just trying to verify correct behavior.

Reply:
What I mean when I say Middleware running on top of each other is them receiving each other's responses, not them invoking each other programmatically and receiving their return values.

Is `finished: true` needed? Can't we just run the Middleware of all the CLI Plugins and the one at which it stops is the last one? Because keep in mind that CLI Plugins aren't aware of each other, so they don't know what will be run on top or below them, so they never know whether they are last. Does that make sense?

In my mind, there basically wouldn't be a difference between a Middleware running on top of a Deployment directly, or on top of Middleware that runs on top of a Deployment. Both could be the case and the Middlewares can decide whether they want to handle both, or just one by looking at the shape of the response of what they're targeting. Things like crash stack traces or logs will be delivered via the response body or meta information (like headers). Logs are already delivered via headers for Lambdas today.

Reply:
Note that I'm not saying we need to make sure that things like logs and crash stack traces are already properly passed around – all I'm saying is that we want to make sure that Middlewares of different CLI Plugins should be able to run on top of each other and communicate via the requests, just like how they communicate with their underlying Deployment if there is no other Middleware underneath them.

Reply:
It's less about knowing whether you're the last middleware to be run, and more about whether the response has been completed.  If a middleware looks like this, I wouldn't expect any other middleware to be run:

```js
export default () => {
  return new Response('hey')
}
```

Reply:
The current implementation would only continue with the next middleware if `x-middleware-next` is present as header, so your Response will indeed not continue, unless you'll add `new Response('hey', { headers: { 'x-middleware-next': '1' }})`.